Fix syscall sysinfo interception on 32 bit (armhf) platforms

https://launchpadlibrarian.net/598840404/buildlog_snap_ubuntu_focal_armhf_lxd-latest-candidate_BUILDING.txt.gz

```
go build -o /build/lxd/parts/lxd/install/bin/lxd -tags=libsqlite3 github.com/lxc/lxd/lxd
# github.com/lxc/lxd/lxd/seccomp
lxd/seccomp/seccomp.go:1738:16: cannot use int64(time.Now().Sub(f.ModTime()).Seconds()) (value of type int64) as type int32 in assignment
lxd/seccomp/seccomp.go:1763:21: cannot use v (variable of type uint64) as type uint32 in assignment
lxd/seccomp/seccomp.go:1765:21: cannot use v (variable of type uint64) as type uint32 in assignment
lxd/seccomp/seccomp.go:1787:18: cannot use uint64(memoryLimit) (value of type uint64) as type uint32 in assignment
lxd/seccomp/seccomp.go:1788:17: invalid operation: info.Totalram - uint64(memoryUsage) (mismatched types uint32 and uint64)
lxd/seccomp/seccomp.go:1808:20: cannot use uint64(swapLimit) (value of type uint64) as type uint32 in assignment
lxd/seccomp/seccomp.go:1809:19: invalid operation: info.Totalswap - uint64(swapUsage) (mismatched types uint32 and uint64)
Failed to run 'override-build': Exit code was 2.
```

--- a/lxd/seccomp/seccomp.go
+++ b/lxd/seccomp/seccomp.go
@@ -1709,6 +1709,7 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 
 	defer l.Debug("Handling sysinfo syscall")
 
+	// Pre-fill sysinfo struct with metrics from host system.
 	info := unix.Sysinfo_t{}
 	err := unix.Sysinfo(&info)
 	if err != nil {
@@ -1717,6 +1718,8 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 
 		return 0
 	}
+
+	instMetrics := Sysinfo{} // Architecture independent place to hold instance metrics.
 
 	cg, err := cgroup.NewFileReadWriter(int(siov.msg.init_pid), liblxc.HasApiExtension("cgroup2"))
 	if err != nil {
@@ -1735,7 +1738,7 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 		return 0
 	}
 
-	info.Uptime = int64(time.Now().Sub(f.ModTime()).Seconds())
+	instMetrics.Uptime = int64(time.Now().Sub(f.ModTime()).Seconds())
 
 	// Get instance process count.
 	pids, err := cg.GetTotalProcesses()
@@ -1746,7 +1749,7 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 		return 0
 	}
 
-	info.Procs = uint16(pids)
+	instMetrics.Procs = uint16(pids)
 
 	// Get instance memory stats.
 	memStats, err := cg.GetMemoryStats()
@@ -1760,9 +1763,9 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 	for k, v := range memStats {
 		switch k {
 		case "shmem":
-			info.Sharedram = v
+			instMetrics.Sharedram = v
 		case "cache":
-			info.Bufferram = v
+			instMetrics.Bufferram = v
 		}
 	}
 
@@ -1784,8 +1787,8 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 		return 0
 	}
 
-	info.Totalram = uint64(memoryLimit)
-	info.Freeram = info.Totalram - uint64(memoryUsage) - info.Bufferram
+	instMetrics.Totalram = uint64(memoryLimit)
+	instMetrics.Freeram = instMetrics.Totalram - uint64(memoryUsage) - instMetrics.Bufferram
 
 	// Get instance swap info.
 	if s.s.OS.CGInfo.Supports(cgroup.MemorySwapUsage, cg) {
@@ -1805,13 +1808,16 @@ func (s *Server) HandleSysinfoSyscall(c Instance, siov *Iovec) int {
 			return 0
 		}
 
-		info.Totalswap = uint64(swapLimit)
-		info.Freeswap = info.Totalswap - uint64(swapUsage)
+		instMetrics.Totalswap = uint64(swapLimit)
+		instMetrics.Freeswap = instMetrics.Totalswap - uint64(swapUsage)
 	}
 
 	// Get writable pointer to buffer of sysinfo syscall result.
 	const sz = int(unsafe.Sizeof(info))
 	var b []byte = (*(*[sz]byte)(unsafe.Pointer(&info)))[:]
+
+	// Write instance metrics to native sysinfo struct.
+	instMetrics.ToNative(&info)
 
 	// Write sysinfo response into buffer.
 	_, err = unix.Pwrite(siov.memFd, b, int64(siov.req.data.args[0]))

--- a/lxd/seccomp/sysinfo.go
+++ b/lxd/seccomp/sysinfo.go
@@ -1,0 +1,13 @@
+package seccomp
+
+// Sysinfo architecture independent sysinfo struct.
+type Sysinfo struct {
+	Uptime    int64
+	Totalram  uint64
+	Freeram   uint64
+	Sharedram uint64
+	Bufferram uint64
+	Totalswap uint64
+	Freeswap  uint64
+	Procs     uint16
+}

--- a/lxd/seccomp/sysinfo_32.go
+++ b/lxd/seccomp/sysinfo_32.go
@@ -1,0 +1,19 @@
+//go:build 386 || arm || ppc || s390 || mips || mipsle
+
+package seccomp
+
+import (
+	"golang.org/x/sys/unix"
+)
+
+// ToNative fills fields from s into native fields.
+func (s *Sysinfo) ToNative(n *unix.Sysinfo_t) {
+	n.Bufferram = uint32(s.Bufferram)
+	n.Freeram = uint32(s.Freeram)
+	n.Freeswap = uint32(s.Freeswap)
+	n.Procs = s.Procs
+	n.Sharedram = uint32(s.Sharedram)
+	n.Totalram = uint32(s.Totalram)
+	n.Totalswap = uint32(s.Totalswap)
+	n.Uptime = int32(s.Uptime)
+}

--- a/lxd/seccomp/sysinfo_64.go
+++ b/lxd/seccomp/sysinfo_64.go
@@ -1,0 +1,19 @@
+//go:build amd64 || ppc64 || ppc64le || arm64 || s390x || mips64 || mips64le || riscv64
+
+package seccomp
+
+import (
+	"golang.org/x/sys/unix"
+)
+
+// ToNative fills fields from s into native fields.
+func (s *Sysinfo) ToNative(n *unix.Sysinfo_t) {
+	n.Bufferram = s.Bufferram
+	n.Freeram = s.Freeram
+	n.Freeswap = s.Freeswap
+	n.Procs = s.Procs
+	n.Sharedram = s.Sharedram
+	n.Totalram = s.Totalram
+	n.Totalswap = s.Totalswap
+	n.Uptime = s.Uptime
+}

